Blog Post: Open Hardware for Open Money (#641)
OpenSats/website: data/blog/open-hardware-for-open-money.mdx

using transparent designs, enable local partners to manufacture and repair
boards, deliver affordable DIY air-gapped signers, and assemble an open mining
stack for independent operators.

These projects are:

- [Bitaxe](#bitaxe)
- [Bitshoka](#bitshoka)
- [Krux](#krux)
- [Krux-Installer](#krux-installer)
- [256 Foundation](#256-foundation)

Having mining hardware and signing devices as public infrastructure lowers
barriers to entry and expands the pool of people who can meaningfully
participate in Bitcoin. Let's take a closer look at how each of these
open-source projects has made an impact over the past year.

[earlier impact reports]: /tags/impact

---

### Bitaxe

[Bitaxe] is an open-source bitcoin mining project, led by hardware designer
[skot], that gives individuals and small operators a practical way to contribute
directly to network security. With [support from OpenSats][bitaxe-grant], the
project's focus has been on closing the gap between open design files and
hardware that people can build, run, and understand. Each release moved that
goal forward by making manufacturable boards with matching firmware available to
anyone willing to assemble them.

> The [BitaxeGamma line] marked a major step in mining decentralization by using

Reviewer: nit, with a space seems better to my eye, though that may require adding the space in a couple further instances below in this document.
```suggestion
The [Bitaxe Gamma line] marked a major step in mining decentralization by using
```

[aditBoard]: https://github.com/skot/aditBoard
[Bitaxe Raw]: https://github.com/bitaxeorg/bitaxe-raw

---

### Krux

[Krux] is a firmware project that turns inexpensive, off-the-shelf circuit
boards into air-gapped bitcoin signing devices. In places like Brazil, taxes and
import fees often double the price of electronics, putting commercial hardware
wallets beyond the reach of many people. [OpenSats funding][krux-grant]
supported two core contributors, [Odudex] and [Tadeubas], who worked to make
self-custody practical on affordable, widely available hardware.

Under Odudex's lead, the [25.03.0] release added support for Taproot and P2WSH
wallets that follow Miniscript. Krux can now load a wallet [descriptor] that
defines who can spend, under which conditions, and when. The device displays it
in a structured way and checks that the cosigners and rules match the user
settings before a transaction is approved. SD-card signing reliability was
improved, so when a transaction file ([PSBT]) moves from one signer to another
on an SD card, Krux is able to keep the information intact, including signatures
already added on other devices, so that multi-signature setups can pass around

> one file and see signatures accumulate. Camera and interface changes in the same

Reviewer 2: ```suggestion -L222+L225
one file and see signatures accumulate. In the same release, the camera gained a zoom option and improved performance under glare, making QR code based signing easier. Interface updates also enhanced keypad layouts and introduced a simpler tamper-check screen optimized for small displays.
```

sensitive data. The Krux Encryption Format (KEF), introduced in the [25.09.0]
release, gives users a way to encrypt seeds and other secrets with protection
against brute force attempts, and the possibility to move those encrypted
envelopes using QR or SD card while staying offline. Contributor [Jean Do] led
the design of KEF as a secure envelope format for air-gapped devices, while
Odudex helped implement the low-level parts into Krux. On top of KEF, the new
Datum tool lets users turn plain text into encrypted records. Users can write
instructions, recovery hints, or account notes and save them in KEF format as QR
codes or SD card files, so important context sits next to wallet backups in a
human-readable yet protected form.

Performance and long-term architecture were also a focus for Odudex. He [reduced
the size of the Krux firmware][25.10.0] by 25% and enabled hardware acceleration
for encryption and hashing on supported boards. Users gained quicker QR
handling, faster transaction signing, and better responsiveness on low-cost
devices. To support this, he wrote cUR, a pure C library for handling UR-based

> QR codes, and integrated it into Krux as a MicroPython module. cUR makes reading

Reviewer 3: Just a nit, could replace "integrated it" by "is integrating it" as I'm still maturing cUR and is not part of releases yet.

need a list of addresses generated from an offline signer.

On the hardware side, Tadeubas used grant funds to bring Krux to the [TZT]
touchscreen device, which combines a milled aluminum case with both touch input
and physical buttons. Porting Krux to TZT required changes in the underlying
[MaixPy] firmware for display, buttons, and camera support, and this work
shipped as new device support in the [v25.10.0][25.10.0] release. TZT gives
users a more rugged and comfortable form factor, which many users prefer for
long-term key storage. Beyond this, Tadeubas is also developing the [Krux Apps]
(Kapps) framework, including a nostr-focused app that would let Krux act as an
offline nostr signer in future versions.

>Projects like Krux don't just add another tool; they strengthen the entire
>Bitcoin self-custody ecosystem by promoting openness, affordability, air-gapped
>security, and resistance to centralized control or suppression. Contributing to
>that effort, with OpenSats' support, has been deeply rewarding.
>
><cite>—Tadeubas, Krux contributor</cite>

Krux runs on widely available K210 boards and now on devices like TZT,

> [Yahboom], [Maix Cube], and [WonderMV], so more users have a practical path to

Reviewer 2: ```suggestion
[Yahboom], [Maix Cube], [WonderMV], and [Embed Fire], so more users have a practical path to
```

[descriptor]: https://bitcoinops.org/en/topics/output-script-descriptors/
[PSBT]: https://bitcoinops.org/en/topics/psbt/
[25.09.0]: https://github.com/selfcustody/krux/releases/tag/v25.09.0
[Jean Do]: https://github.com/jdlcdl
[25.10.0]: https://github.com/selfcustody/krux/releases/tag/v25.10.0
[Kern]: https://github.com/odudex/Kern
[vulnerability]: https://github.com/selfcustody/krux/discussions/472
[anti-rollback protection]: https://github.com/selfcustody/krux/pull/583
[Craig Watkins]: https://github.com/bitcoin-dot-org/Bitcoin.org/pull/4364#issuecomment-2828998014
[Embit library]: https://embit.rocks/#/
[groups]: https://github.com/selfcustody/krux/pull/541
[highlighting]: https://github.com/selfcustody/krux/pull/629
[export wallet addresses to SD]: https://github.com/selfcustody/krux/pull/638
[TZT]: https://selfcustody.github.io/krux/parts/#tzt
[MaixPy]: https://github.com/selfcustody/MaixPy
[Krux Apps]: https://github.com/selfcustody/krux/pull/485
[Yahboom]: https://selfcustody.github.io/krux/parts/#yahboom-k210-module
[Maix Cube]: https://selfcustody.github.io/krux/parts/#maix-cube
[WonderMV]: https://selfcustody.github.io/krux/parts/#wondermv

> 

Reviewer 2: ```suggestion
[Embed Fire]: https://selfcustody.github.io/krux/parts/#embed-fire

```

tamper-check screen optimized for small displays.

A second set of changes from Odudex focused on how Krux protects and stores
sensitive data. The Krux Encryption Format (KEF), introduced in the [25.09.0]
release, gives users a way to encrypt seeds and other secrets with protection
against brute force attempts, and the possibility to move those encrypted
envelopes using QR or SD card while staying offline. Contributor [Jean Do] led
the design of KEF as a secure envelope format for air-gapped devices, while
Odudex helped implement the low-level parts into Krux. On top of KEF, the new
Datum tool lets users turn plain text into encrypted records. Users can write
instructions, recovery hints, or account notes and save them in KEF format as QR
codes or SD card files, so important context sits next to wallet backups in a
human-readable yet protected form.

Performance and long-term architecture were also a focus for Odudex. He [reduced
the size of the Krux firmware][25.10.0] by 25% and enabled hardware acceleration
for encryption and hashing on supported boards. Users gained quicker QR
handling, faster transaction signing, and better responsiveness on low-cost

> devices. To support this, he wrote cUR, a pure C library for handling UR-based

Reviewer 4: Maybe adding link to cUR: https://github.com/odudex/cUR

Odudex has also been building [Kern], a C implementation of a Krux-style signer
for the ESP32-P4 chip. Kern already runs on several boards, can scan static and
animated QR codes, load keys, and sign transactions, and aims to deliver
reliable air-gapped signers for under fifty dollars on widely available
hardware.

>By building Kern on widely available, future-proof, low-cost hardware without
>sacrificing performance or usability, we can bring reliable Bitcoin
>self-custody to people who were previously priced out of quality signing
>devices. The tools I created along the way, like cUR, are open for anyone to
>use, and hopefully can benefit other projects.
>
><cite>—Odudex, Krux contributor</cite>

Working alongside Odudex, Tadeubas concentrated on safety checks, address
handling, security hardening, and device support in the main Krux firmware. He
identified and fixed a [vulnerability] that allowed firmware to be compromised
through the SD card, and later implemented [anti-rollback protection] after
[Craig Watkins] raised the issue during the review process to list Krux on
Bitcoin.org. His work on the [Embit library] and Krux itself improved how the

> signer understands complex wallet definitions. Krux is now is able to recognize

Reviewer 4: ```suggestion
signer understands complex wallet definitions. Krux is now able to recognize
```

---
title: "Open Hardware for Open Money"

> date: '2026-04-20'

Reviewer 5: ```suggestion
date: '2026-04-21'
```
